### PR TITLE
Sector storage checkPrecommit (checks #2)

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Add CheckPrecommit()
Add IpldStore wrapper for API

### CODE REVIEW

File: core/api/make.cpp [hunk 1]
```diff
@@ -116,21 +116,34 @@ namespace fc::api {
...
-      OUTCOME_TRY(seal_type,
-                  primitives::sector::sealProofTypeFromSectorSize(
-                      state.info.sector_size));
-      OUTCOME_TRY(win_type,
-                  primitives::sector::getRegisteredWinningPoStProof(seal_type));
-      OUTCOME_TRY(state.visitProvingSet([&](auto id, auto &info) {
-        sectors.push_back({win_type, id, info.info.sealed_cid});
-      }));
-      if (!sectors.empty()) {
+      RleBitset sectors1;
```
Comment (artyom-yurin) [inline core/api/make.cpp:119]: It might be worth choosing a more informative name

File: core/api/make.cpp [hunk 2]
```diff
@@ -116,21 +116,34 @@ namespace fc::api {
...
+            primitives::sector::sealProofTypeFromSectorSize(minfo.sector_size));
+        OUTCOME_TRY(
+            win_type,
+            primitives::sector::getRegisteredWinningPoStProof(seal_type));
         OUTCOME_TRY(indices,
                     proofs::Proofs::generateWinningPoStSectorChallenge(
-                        win_type, miner.getId(), post_rand, sectors.size()));
-        std::vector<SectorInfo> result;
+                        win_type, miner.getId(), post_rand, sectors1.size()));
+        std::vector<uint64_t> sectors2{sectors1.begin(), sectors1.end()};
```
Comment (artyom-yurin) [inline core/api/make.cpp:143]: same naming issue

File: core/miner/windowpost.cpp [hunk 3]
```diff
@@ -54,34 +54,54 @@ namespace fc {
...
-      auto parts{deadlines.count(part_size, deadline.index).first};
-      if (parts != 0) {
-        using vm::actor::builtin::miner::SubmitWindowedPoSt;
-        SubmitWindowedPoSt::Params params;
-        auto first_part{deadlines.partitions(part_size, deadline.index).first};
-        for (auto i{0u}; i < parts; ++i) {
-          params.partitions.push_back(first_part + i);
-        }
+
+      // TODO: check recoveries and faults
```
Comment (artyom-yurin) [inline core/miner/windowpost.cpp:58]: JIRA Label

File: core/sector_storage/checks/checks.cpp [hunk 4]
```diff
@@ -46,6 +65,87 @@ namespace fc::sector_storage::checks {
...
+      return ChecksError::kExpiredTicket;
+    }
+
+    OUTCOME_TRY(
+        state_sector_precommit_info,
+        getStateSectorPreCommitInfo(address, sector_info, tipset_key, api));
+    if (state_sector_precommit_info.has_value()
+        && (state_sector_precommit_info.value().info.seal_epoch
+            != sector_info.ticket_epoch)) {
+      return ChecksError::kBadTicketEpoch;
```
Comment (artyom-yurin) [inline core/sector_storage/checks/checks.cpp:143]: No check for "already in chain"

**Review by artyom-yurin — APPROVED** [review]

File: core/sector_storage/checks/checks.cpp [hunk 5]
```diff
@@ -46,6 +62,89 @@ namespace fc::sector_storage::checks {
...
+
+  outcome::result<boost::optional<SectorPreCommitOnChainInfo>>
+  getStateSectorPreCommitInfo(const Address &miner_address,
+                              const SectorInfo &sector_info,
+                              const TipsetKey &tipset_key,
+                              const std::shared_ptr<Api> &api) {
+    boost::optional<SectorPreCommitOnChainInfo> result;
+
+    OUTCOME_TRY(actor, api->StateGetActor(miner_address, tipset_key));
+    ApiIpfsDatastore ipfs{api};
```
Comment (turuslan) [inline core/sector_storage/checks/checks.cpp:102]: make_shared

**Review by turuslan — APPROVED** [review]

**Pull request merged**